### PR TITLE
[skip-changelog] Avoid getting locked up in a perpetual symlink loop when loading a library

### PR DESCRIPTION
**Pull request by MatteoPologruto**
## Please check if the PR fulfills these requirements

See [how to contribute](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/)

- [x] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-cli/pulls)
      before creating one)
- [x] The PR follows
      [our contributing guidelines](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#pull-requests)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] `UPGRADING.md` has been updated with a migration guide (for breaking changes)
- [ ] `configuration.schema.json` updated if new parameters are added.

## What kind of change does this PR introduce?
Code imperfection fix

## What is the current behavior?
When loading a library or sketch that contains multiple symlinks to a parent path, Arduino CLI locks up in a perpetual loop.

## What is the new behavior?
Arduino CLI fails with an error when a symlink loop is encountered.

## Does this PR introduce a breaking change, and is [titled accordingly](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#breaking)?
No

### CODE REVIEW

**Review by per1234 — APPROVED** [review]
Thanks Matteo!

File: arduino/libraries/libraries_test.go [hunk 1]
```diff
@@ -85,3 +88,89 @@ func TestLibrariesLoader(t *testing.T) {
...
+	examplesPath := libraryPath.Join("examples")
+	err = examplesPath.Mkdir()
+	require.NoError(t, err)
+
+	// It's probably most friendly for contributors using Windows to create the symlinks needed for the test on demand.
+	err = os.Symlink(examplesPath.Join("..").String(), examplesPath.Join("UpGoer1").String())
+	require.NoError(t, err, "This test must be run as administrator on Windows to have symlink creation privilege.")
+	// It's necessary to have multiple symlinks to a parent directory to create the loop.
+	err = os.Symlink(examplesPath.Join("..").String(), examplesPath.Join("UpGoer2").String())
+	require.NoError(t, err)
```
Comment (cmaglie) [inline arduino/libraries/libraries_test.go:151]: Can't we save the test library directly in `testdata` directly in git instead of re-creating it?

File: arduino/libraries/libraries_test.go [hunk 2]
```diff
@@ -85,3 +88,89 @@ func TestLibrariesLoader(t *testing.T) {
...
+				return true
+			default:
+				return false
+			}
+		},
+		20*time.Second,
+		10*time.Millisecond,
+		"Infinite symlink loop while loading library",
+	)
+	require.True(t, noHang)
```
Comment (cmaglie) [inline arduino/libraries/libraries_test.go:131]: IIUC this assert will fail after 20 seconds, that seems way too much, we can assume that `Load` should complete in a few milliseconds. I'd suggest to use 2 seconds instead.
Also the check may be done much more easily like this:
```suggestion
	select {
	case <-done:
	case <-time.After(2*time.Second):
		require.FailNow(t, "Load didn't complete in the allocated time.")
	}
```

File: arduino/libraries/loader.go [hunk 3]
```diff
@@ -184,7 +184,9 @@ func addExamplesToPathList(examplesPath *paths.Path, list *paths.PathList) error
 		_, err := sketch.New(file)
 		if err == nil {
 			list.Add(file)
-		} else if file.IsDir() {
+		} else if isDir, err := file.IsDirCheck(); err != nil {
+			return err
+		} else if isDir {
```
Comment (cmaglie) [inline arduino/libraries/loader.go:189]: I don't understand why this change fixes the issue, can you explain it a bit?
Is `file.IsDir()` returning `true` on symlinks? why `IsDirCheck` would return error instead? 🤔

**Review by cmaglie — CHANGES_REQUESTED** [review]

File: arduino/libraries/libraries_test.go (hunk 1 above)
Comment (per1234) [reply arduino/libraries/libraries_test.go:151]: The reason for generating the symlinks on the fly is that, by default, symlinks are only supported on Windows when in admin mode. So if a Windows user clones a repo that contains symlinks when not in admin mode, the symlinks are not generated (you only get a file that contains the target path as text instead of the link).

For this reason, I think it is more contributor friendly to generate symlinks on the fly. The contributor still needs to run the unit tests as administrator, but the test will clearly communicate the requirement to them if they try to run it without doing so.

But that is only about the symlinks. The symlinks could be added on the fly to a library from testdata instead of generating the whole thing.

**Review by umbynos — APPROVED** [review]

**Review by cmaglie — APPROVED** [review]
